src/programs.c

```diff
@@ -637,5 +776,921 @@ int uprobe__read_return_string(struct pt_regs *ctx)
+    else if ((void *) ptr == NULL)                                              \
+    {                                                                           \
+        goto Next;                                                              \
+    }                                                                           \
+    else                                                                        \
+    {                                                                           \
+        count = bpf_probe_read_str(&ev->u.v.value, VALUE_SIZE, (void *) ptr);   \
+        if (count == VALUE_SIZE) {                                              \
+            ev->u.v.truncated = TRUE;                                           \
+        }                                                                       \
+        if (count <= 0) {                                                       \
+            goto Next;                                                          \
+        }                                                                       \
+    }                                                                           \
+    push_telemetry_event(ev);                                                   \
+    ii++;
+
+#define READ_LOOP_N(PTR, N) REPEAT_##N(READ_LOOP(PTR);)
+
+static __always_inline int enter_exec(syscall_pattern_type_t sp, int fd,
+                                      const char __user *filename,
+                                      const char __user *const __user *argv,
+                                      const char __user *const __user *envp,
+                                      int flags, struct pt_regs *ctx, u32 ppid,
+                                      u32 luid)
+{
+    u64 id = 0; // reuse ID to save space
+    ptelemetry_event_t pev = bpf_map_lookup_elem(&telemetry_stack, (u32 *) &id);
+    if (!pev) // this should never happen, but the verifier complains otherwise
+    {
+        goto Exit;
+    }
+    telemetry_event_t sev = {0};
+    __builtin_memcpy(&sev, pev, sizeof(telemetry_event_t));
+    ptelemetry_event_t ev = &sev;
+
+    id = bpf_get_prandom_u32();
+    ev->id = id;
+    ev->done = FALSE;
+    ev->telemetry_type = 0,
+    ev->u.v.value[0] = '\0';
+    ev->u.v.truncated = FALSE;
+
+    FILL_TELEMETRY_SYSCALL_EVENT(ev, sp);
+    u32 pid = pid_tgid >> 32;
+    ev->u.syscall_info.ppid = ppid;
+    ev->u.syscall_info.luid = luid;
+    push_telemetry_event(ev);
+
+    bpf_map_update_elem(&telemetry_ids, &pid, &id, BPF_ANY);
+
+
+    // explicit null check to satisfy the verifier
+    if (!filename)
+        goto Exit;
+
+    ev->id = id;
+    ev->done = FALSE;
+    ev->telemetry_type = TE_EXE_PATH;
+    ev->u.v.truncated = FALSE;
+    long count = 0;
+    count = bpf_probe_read_str(&ev->u.v.value, VALUE_SIZE, (void*) filename);
+    if (count == VALUE_SIZE) {
+        ev->u.v.truncated = TRUE;
+    }
+    push_telemetry_event(ev);
+
+
+    bpf_tail_call(ctx, &tail_call_table, 0);
```

Comment:
I'm thinking we should probably create macros or an enum to represent the different tail cail slots, thoughts?

Reply:
Yeah, that'll make it less error prone on the `oxidebpf` side.